contributors/design-proposals/ipvs-proxy.md

```diff
@@ -0,0 +1,152 @@
+# Alpha Version IPVS Load Balancing Mode in Kubernetes
+
+## Summary
+
+There’s been a lot discussion and voice of enabling IPVS as in-cluster service load balancing mode.  IPVS performs better than iptables, meanwhile supports more sophisticated load balancing algorithms than iptables (least load, least connections, locality, weighted) as well as other useful features (e.g. health checking, retries etc).
+
+This page summarizes what’s expected in alpha version IPVS load balancing support, includes Kubernetes user behavior changes, build and deployment changes, and the test validation planned. 
+
+##  Kubernetes behavior change
+
+### Changes to kube-proxy startup parameter
+
+In addition to existing userspace and iptables mode, ipvs mode is configured via `--proxy-mode=ipvs`. In alpha version, it implicitly uses IPVS NAT(masq) mode.
+
+A new kube-proxy parameter `ipvs-scheduler` will be added to specify IPVS load balancing algorithm, the parameter is --ipvs-scheduler.  Below is list of supported values, if it’s not configured `rr` is default value, if it’s incorrectly configured kube-proxy will exit with error message.
+
+- rr: round-robin
+
+
+- lc: least connection
+
+
+- dh: destination hashing
+
+
+- sh: source hashing
+
+
+- sed: shortest expected delay
+
+
+- nq: never queue
+
+For more details about it, refer to [http://kb.linuxvirtualserver.org/wiki/Ipvsadm](http://kb.linuxvirtualserver.org/wiki/Ipvsadm)
+
+In order to clean up inactive rules(includes iptables rules and ipvs rules), we will introduce a new  kube-proxy parameter `--cleanup-proxyrules ` and mark the older `--cleanup-iptables` deprecated. Unfortunately, there is no way to distinguish whether an ipvs service is created by ipvs proxy or other process, `--cleanup-proxyrules` will clear all ipvs service in a host.
```

Comment:
Can you only clean rules with RC1918 addresses?
[m1093782566] It still has a possibility of clearing ipvs rules created by other process although the the possibility is low.
[m1093782566] Kube-apiserver knows the service cluster IP range through its--service-cluster-ip-range parameter. However, kube-proxy knows nothing about that. I don't suggest to add a new --service-cluster-ip-range flag in kube-proxy since it will easily conflicts with kube-apiserver's flag. Even though kube-proxy clearing all ipvs rules in the service cluster IP range, it may leave some ipvs rules in external IP, exteral LB ingress IP and Node IP.

Clearing all ipvs rules with RFC1918 addresses probably is easier to implement and has lower possibility of clearing user's existing ipvs rules. 

@fisherxu What do you think about it?